metriken/src/histogram.rs

```diff
@@ -0,0 +1,143 @@
+use crate::{Metric, Value};
+use parking_lot::RwLock;
+
+use std::sync::OnceLock;
+
+pub use histogram::{Bucket, Config, Error, Snapshot};
+
+/// A histogram that uses free-running atomic counters to track the distribution
+/// of values. They are only useful for recording values and producing
+/// [`crate::Snapshot`]s of the histogram state which can then be used for
+/// reporting.
+///
+/// The `AtomicHistogram` should be preferred when individual events are being
+/// recorded. The `RwLockHistogram` should be preferred when bulk-updating the
+/// histogram from pre-aggregated data with a compatible layout.
+pub struct AtomicHistogram {
+    inner: OnceLock<histogram::AtomicHistogram>,
+    config: Config,
+}
+
+impl AtomicHistogram {
+    /// Create a new [`::histogram::AtomicHistogram`] with the given parameters.
+    ///
+    /// # Panics
+    /// This will panic if the `grouping_power` and `max_value_power` do not
+    /// adhere to the following constraints:
+    ///
+    /// - `max_value_power` must be in the range 1..=64
+    /// - `grouping_power` must be in the range `0..=(max_value_power - 1)`
+    pub const fn new(grouping_power: u8, max_value_power: u8) -> Self {
+        let config = match ::histogram::Config::new(grouping_power, max_value_power) {
+            Ok(c) => c,
+            Err(_) => panic!("invalid histogram config"),
+        };
+
+        Self {
+            inner: OnceLock::new(),
+            config,
+        }
+    }
+
+    /// Increments the bucket for a corresponding value.
+    pub fn increment(&self, value: u64) -> Result<(), Error> {
+        self.get_or_init().increment(value)
+    }
+
+    pub fn config(&self) -> Config {
+        self.config
+    }
+
+    /// Create a new snapshot from the histogram.
+    pub fn snapshot(&self) -> Option<Snapshot> {
```

Comment:
Curious about the `Option` here -- it seems to me like `histogram.snapshot()` should never fail? Or are there conditions where it would return `None`?

Reply:
The histogram might not be initialized when we call snapshot. Currently it is only initialized when we first record into it. An unused (in terms of write path) histogram will never be initialized.

Reply:
Ah, makes sense, it's a `get` not a `get_or_init`, thanks.